### DISCUSSION TITLE
Incorrect .rs files interpretation in language activity

### DISCUSSION THREAD

**Discussion by Pandicon** [lowlighter/metrics#1107] [Q&A]
I am currently mainly working in Rust and I was wondering why the "Language activity" section was saying I was using RenderScript (see screenshot). I recenly looked into it and found out that both Rust and RenderScript use the .rs file extension. Both the "Most used languages" and "Recently used languages" sections interpret them correctly as Rust source files.
I tried to alias RenderScript to Rust via `plugin_languages_aliases: RenderScript:Rust`, but it didn't affect the "Language activity" section so I undid it to not confuse me later. I am using the GitHub actions version using `lowlighter/metrics@latest`.
Is there a way to fix it/at least alias it? The files include all sorts of Rust things (`pub`, `fn`, `use`...) and are usually at least 100 lines long, so it shouldn't be an issue of not enough data.
![language_interpretation](https://user-images.githubusercontent.com/70060103/174797746-0685191e-65dd-4fa0-abdc-0ef3e5d6a700.png)

**Comment by spenserblack** [comment]
Probably related to https://github.com/lowlighter/metrics/discussions/1047 and https://github.com/lowlighter/metrics/issues/551.

Basically, like you said, RenderScript and Rust share the same extension, `.rs`. When metrics (via linguist-js) can't tell which language it is, it defaults to RenderScript. This happens because metrics breaks down your commits into patch files.
https://github.com/lowlighter/metrics/blob/552c2a245fb3b7880c1e198a1a67bede5c9953df/source/plugins/habits/index.mjs#L42-L55
Even if the file as a whole contains enough Rust-like code to be picked up by heuristics, it's likely that the patches won't. This might relate to your commit style, too: big monolithic commits might result in more accurate language metrics than small, atomic commits (though the latter is much better IMO).

**Reply by Pandicon** [reply]
I saw the issue #551 and the solution there was to alias RenderScript as Rust (but it was only for the "Recently used languages" section). Is there no workaround for "Language activity"? I tried to set [.rs files to be treated as Rust](https://github.com/github/linguist/blob/master/docs/overrides.md#using-gitattributes) files by Github Linguist by putting `*.rs linguist-language=Rust` into the `.gitattributes` file in the repo I am active in, but it didn't change anything.

**Reply by spenserblack** [reply]
The analyzer is supposed to fetch your `.gitattributes`, so that might be a bug.
https://github.com/lowlighter/metrics/blob/552c2a245fb3b7880c1e198a1a67bede5c9953df/source/plugins/habits/index.mjs#L2
https://github.com/lowlighter/metrics/blob/552c2a245fb3b7880c1e198a1a67bede5c9953df/source/plugins/habits/index.mjs#L106
https://github.com/lowlighter/metrics/blob/552c2a245fb3b7880c1e198a1a67bede5c9953df/source/plugins/languages/analyzers.mjs#L165

**Reply by lowlighter** [reply]
Not sure why the `.gitattributes` isn't loaded correctly 😕 

But I think one way to solve this issue may be to fetch the whole file to perform github-linguist analysis on it, and only count lines/bytes from the patch itself so in theory the language should be detected more accurately with more context while the stats remain the same

**Reply by Pandicon** [reply]
Even after I pushed a 4.77kB file with multiple struct definitions and impl blocks for them, nothing was reflected in the stats. I don't know how the patches work, but I suppose it is the added/removed highlighting GitHub shows. If that is the case, in the commit the patch should contain the whole file, including a lot of clearly Rust stuff (use, struct, impl, ...). Not sure how helpful this will be, but if I understand how the patches work, I feel like fetching the whole file may not really help.

**Reply by spenserblack** [reply]
@lowlighter The patch's lines start with `-` or `+`, right?

@Pandicon That can be why `use`, `impl`, etc. aren't picked up by the heuristics.
The [current regex](https://github.com/github/linguist/blob/6a3bd9faf83bc58c91522f9e0ff6c3d645545661/lib/linguist/heuristics.yml#L561) matches on a line like `use`, but not a line like `+use`.

**Reply by lowlighter** [reply]
> The patch's lines start with `-` or `+`, right?

Yeah, patches looks like this:
![image](https://user-images.githubusercontent.com/22963968/176345746-b2e695c6-5a24-47ca-aa69-ca92569c504a.png)

(e.g. [`3d1d5ed`](https://api.github.com/repos/lowlighter/metrics/commits/3d1d5ed37eb0ea1e33b2b176654c28cc91d26138))

It's basically what it output by a `git diff`

The `+` part is supposed to be trimmed by this line:
https://github.com/lowlighter/metrics/blob/8e381688153bad0c08e42d8498baee93f1b8d6ed/source/plugins/languages/analyzers.mjs#L143

But added lines are not trimmed (it just remove the `+` char), so that'd why it probably doesn't match heuristics since they have the `^` in the regex with no whitespaces
